nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/bin/Bin.java

```diff
@@ -173,4 +174,19 @@ public long getBinAge() {
         final long ageInNanos = System.nanoTime() - creationMomentEpochNs;
         return TimeUnit.MILLISECONDS.convert(ageInNanos, TimeUnit.NANOSECONDS);
     }
+
+    /**
+     * @return the UUID of the flow file generated by the merge of this bin's content
+     */
+    public String getUuid() {
+        return uuid;
+    }
+
+    /**
+     * Set the UUID of the flow file generated by the merge of this bin's content
+     * @param uuid UUID to set
+     */
+    public void setUuid(String uuid) {
```

Comment:
Is this the only place the association can be done? Would be nicer if the UUID could be added at transfer time rather than extending the Bin API.

Reply:
I'm a bit concerned about this approach as well... As is, in this PR, the abstract class is assuming that this UUID has been set by the concrete implementation. If it is not, then it will throw a NPE, I do believe, from looking at the code. We definitely need to address that. I do believe that a cleaner way may be possible, but off the top of my head i don't have one. I'd have to look at the code more. I'll be out of town for the next week, without my computer. But I'd be happy to look into this a bit more when I can, if you guys haven't resolved this by that time.

Reply:
After reviewing this again, I think we've got 2 approaches that I would recommend. My preferred approach would be to update the `boolean processBin(Bin, ProcessContext)` method to return a `BinProcessingResult` object, where `BinProcessingResult` would be a new class that has the boolean of whether or not the bin was committed and also has a Map<String, String> of attributes that should be added to the 'original' FlowFiles. This is more flexible and allows us to make other modifications more cleanly in the future. This is not considered part of our API but rather is just a convenience class, so we can change it in any minor version of NiFi.

The second option would be to keep as-is, but at least just check if the bin uuid is null. If so, just ignore it instead of trying to add it to the FlowFiles. This option, though, I feel is a little less clean and requires that implementations know that this 'magical' UUID exists on the bin. It's also less flexible for future updates.